# IS 362 - Project 4
Sunah Lee

## Predictive Analysis using scikit-learn¶

Your assignment is to:

• Start with the mushroom data in the pandas DataFrame that you constructed in your “Assignment – Preprocessing Data with sci-kit learn.”

• Use scikit-learn to determine which of the two predictor columns that you selected (odor and one other column of your choice) most accurately predicts whether or not a mushroom is poisonous. There is an additional challenge here—to use scikit-learn’s predictive classifiers, you’ll want to convert each of your two (numeric categorical) predictor columns into a set of columns. See for one approach pandas get_dummies() method.

• Clearly state your conclusions along with any recommendations for further analysis.

This is by design a very open-ended assignment. You’re encouraged to go through the resources in your Week 12 folder. In particular, if you understand the process used in the Kevin Markham videos on Machine Learning with scikit-learn to predict iris species from four predictor variables, you should be able to transfer your learnings to complete this task.

Your deliverable is a Jupyter Notebook that performs these predictive analysis tasks.
If you chose a different dataset for the preliminary processing assignment, you may continue to use that dataset here.
You should post the Jupyter Notebook (.ipynb) file in your GitHub repository, and provide the appropriate URL to your GitHub repository in your assignment link. You should also have the original data file accessible through your code—for example, read directly from the UCI repository or stored in a GitHub repository.

### Step 1 - Read the data and tidy up the columns.

In [6]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
%matplotlib inline
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn import linear_model

shroomdata = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/mushroom/agaricus-lepiota.data',sep=',', header=None, usecols=[0,3,5,9], names=["Edible","Cap Color","Odor","Gill Color"])
shroomdata.replace(to_replace={"Edible":{'p': 1, 'e': 0}}, inplace=True)
shroomdata.replace(to_replace={"Cap Color":{'n':0, 'b':1, 'c':2, 'g':3, 'r':4, 'p':5, 'u':6, 'e':7, 'w':8, 'y':9}}, inplace=True)
shroomdata.replace(to_replace={"Odor":{'a':0, 'l':1, 'c':2, 'y':3, 'f':4, 'm':5, 'n':6, 'p':7, 's':8}}, inplace=True)
shroomdata.replace(to_replace={"Gill Color":{'k':0, 'n':1, 'b':2, 'h':3, 'g':4, 'r':5, 'o':6, 'p':7, 'u':8, 'e':9, 'w':10, 'y':11}}, inplace=True)

shroomdata.head(10)

Edible  Cap Color  Odor  Gill Color
0       1          0     7           0
1       0          9     0           0
2       0          8     1           1
3       1          8     7           1
4       0          3     6           0
5       0          9     0           1
6       0          8     0           4
7       0          8     1           1
8       1          8     7           7
9       0          9     0           4

In [2]:
shroomdata.groupby('Edible').count()

Cap Color  Odor  Gill Color
Edible                             
0            4208  4208        4208
1            3916  3916        3916

In the database, 4,208 mushrooms are edible and 3,916 mushrooms are poisonous.

### Step 2 - Determine which two traits (Odor & Cap Color, or Order & Gill Color is the best predictor of poisonous mushroom.

#### I. Review mushrooms by Odor

In [7]:
shroomdata.groupby(['Edible', 'Odor']).count()

Cap Color  Gill Color
Edible Odor                       
0      0           400         400
       1           400         400
       6          3408        3408
1      2           192         192
       3           576         576
       4          2160        2160
       5            36          36
       6           120         120
       7           256         256
       8           576         576

The table above show that edible mushrooms have can be grouped into 3 odors. 
But order 6 can be found in both edible and poisonous mushrooms.

#### II. Review mushrooms by Cap Color

In [4]:
shroomdata.groupby(['Edible', 'Cap Color']).count()

Odor  Gill Color
Edible Cap Color                  
0      0          1264        1264
       1            48          48
       2            32          32
       3          1032        1032
       4            16          16
       5            56          56
       6            16          16
       7           624         624
       8           720         720
       9           400         400
1      0          1020        1020
       1           120         120
       2            12          12
       3           808         808
       5            88          88
       7           876         876
       8           320         320
       9           672         672

Per table above, mushrooms with a Cap Color 4 or 6 are edible and these cap colors cannot be found in poisonous mushrooms.

#### III. Review by Gill Color

In [5]:
shroomdata.groupby(['Edible', 'Gill Color']).count()

Cap Color  Odor
Edible Gill Color                 
0      0                 344   344
       1                 936   936
       3                 204   204
       4                 248   248
       6                  64    64
       7                 852   852
       8                 444   444
       9                  96    96
       10                956   956
       11                 64    64
1      0                  64    64
       1                 112   112
       2                1728  1728
       3                 528   528
       4                 504   504
       5                  24    24
       7                 640   640
       8                  48    48
       10                246   246
       11                 22    22

## Conclusion

The analysis from this project reveal that odor is the most accurate predictor of edible or poisonous mushroom. Although there is room for error when using either cap color or gill color, gill color seems to be the better predictor.